From 668abf9dadd4ece0b27017149a0139156871aac4 Mon Sep 17 00:00:00 2001
From: Daniel Lenski <dlenski@amazon.com>
Date: Wed, 14 Jun 2023 11:38:19 -0700
Subject: [PATCH 1/7] [MDEV-28634] Do not allow silent downgrade of connections
 where SSL is requested

The 2015 commit
https://github.com/mariadb-corporation/mariadb-connector-c/commit/23895fbd4#diff-4339ae6506ef1fb201f6f836085257e72c191d2b4498df507d499fc30d891005
was described as "Fixed gnutls support", which is an extremely incomplete
description of what it actually does.

In that commit, the Connector/C client authentication plugin was modified to
abort following the initial server greeting packet if the client has
requested a secure transport (`mariadb --ssl`), but the server does not
advertise support for SSL/TLS.

However, there's a crucial caveat here:

If the client has requested secure transport (`mariadb --ssl`) but has not
requested verification of the server's TLS certificate
(`mariadb --ssl --ssl-verify-server-cert`), then the client will silently
accept an unencrypted connection, without even printing a diagnostic message.

Thus, any such client is susceptible to a trivial downgrade to an
unencrypted connection by an on-path attacker who simply flips the TLS/SSL
capability bit in the advertised server capabilities in the greeting packet.

The entire design of MariaDB's TLS support is severely flawed in terms of
user expectations surrounding secure-by-default connections: the `--ssl`
option SHOULD imply `--ssl-verify-server-cert` BY DEFAULT; if a client
actually wants a TLS connection that's susceptible to a trivial MITM'ed
by any pervasive or on-path attacker, that should be an exceptional case
(e.g. `--insecure-ssl-without-server-cert-verification`) rather than the
default.

Even without resolving the issue of no default verification of server
certificates, the issue of silent downgrade to unencrypted connections can
and should be resolved.

Backwards compatibility: This is an INTENTIONAL backwards-incompatible
change to prevent clients from being silently downgraded to unencrypted
connections, when they've specified `--ssl` and thus clearly indicated that
they do not want unencrypted connections.

All new code of the whole pull request, including one or several files
that are either new files or modified ones, are contributed under the
BSD-new license. I am contributing on behalf of my employer Amazon Web
Services, Inc.
---
 plugins/auth/my_auth.c | 17 +++++------------
 1 file changed, 5 insertions(+), 12 deletions(-)

diff --git a/plugins/auth/my_auth.c b/plugins/auth/my_auth.c
index faea968c5..cf8a7a365 100644
--- a/plugins/auth/my_auth.c
+++ b/plugins/auth/my_auth.c
@@ -277,22 +277,15 @@ static int send_client_reply_packet(MCPVIO_EXT *mpvio,
     mysql->options.extension->tls_allow_invalid_server_cert= 1;
   }
 
-  /* if server doesn't support SSL and verification of server certificate
-     was set to mandatory, we need to return an error */
+  /* if server doesn't support SSL, we need to return an error */
   if (mysql->options.use_ssl && !(mysql->server_capabilities & CLIENT_SSL))
   {
-    if (!mysql->options.extension->tls_allow_invalid_server_cert ||
-        mysql->options.extension->tls_fp ||
-        mysql->options.extension->tls_fp_list)
-    {
-      my_set_error(mysql, CR_SSL_CONNECTION_ERROR, SQLSTATE_UNKNOWN,
-                          ER(CR_SSL_CONNECTION_ERROR), 
-                          "SSL is required, but the server does not support it");
-      goto error;
-    }
+    my_set_error(mysql, CR_SSL_CONNECTION_ERROR, SQLSTATE_UNKNOWN,
+                        ER(CR_SSL_CONNECTION_ERROR),
+                        "SSL is required, but the server does not support it");
+    goto error;
   }
 
-
   /* Remove options that server doesn't support */
   mysql->client_flag= mysql->client_flag &
                        (~(CLIENT_COMPRESS | CLIENT_ZSTD_COMPRESSION | CLIENT_SSL | CLIENT_PROTOCOL_41) 

From 71b52a066c8654ab6796b5a2e98c1d7a46e540f3 Mon Sep 17 00:00:00 2001
From: Daniel Lenski <dlenski@amazon.com>
Date: Wed, 14 Jun 2023 12:25:39 -0700
Subject: [PATCH 2/7] [MDEV-28634] Move check for server TLS/SSL capability to
 mthd_my_real_connect

Two reasons:

1. Reduction of attack surface

   As soon as the client receives the server's capability flags, it knows
   whether the server supports TLS/SSL.

   If the server does not support TLS/SSL, but the client expects and
   requires it, the client should immediately abort at this point in order
   to truncate any code paths by which it could inadvertently continue to
   communicate without TLS/SSL.

2. Separation of concerns

   Whether or not the server supports TLS/SSL encryption at the transport
   layer (TLS stands for TRANSPORT-layer security) is a logically separate
   issue from what APPLICATION-layer authentication modes the client and
   server support or should use.

All new code of the whole pull request, including one or several files
that are either new files or modified ones, are contributed under the
BSD-new license. I am contributing on behalf of my employer Amazon Web
Services, Inc.
---
 libmariadb/mariadb_lib.c | 10 ++++++++++
 plugins/auth/my_auth.c   |  9 ---------
 2 files changed, 10 insertions(+), 9 deletions(-)

diff --git a/libmariadb/mariadb_lib.c b/libmariadb/mariadb_lib.c
index 52f968894..7bff96912 100644
--- a/libmariadb/mariadb_lib.c
+++ b/libmariadb/mariadb_lib.c
@@ -1943,6 +1943,16 @@ MYSQL *mthd_my_real_connect(MYSQL *mysql, const char *host, const char *user,
     }
   }
 
+  /* We now know the server's capabilities. If the client wants TLS/SSL,
+   * but the server doesn't support it, we should immediately abort.
+   */
+  if (mysql->options.use_ssl && !(mysql->server_capabilities & CLIENT_SSL))
+  {
+    SET_CLIENT_ERROR(mysql, CR_SSL_CONNECTION_ERROR, SQLSTATE_UNKNOWN,
+                     "Client requires TLS/SSL, but the server does not support it");
+    goto error;
+  }
+
   /* Set character set */
   if (mysql->options.charset_name)
     mysql->charset= mysql_find_charset_name(mysql->options.charset_name);
diff --git a/plugins/auth/my_auth.c b/plugins/auth/my_auth.c
index cf8a7a365..34bdfa29b 100644
--- a/plugins/auth/my_auth.c
+++ b/plugins/auth/my_auth.c
@@ -277,15 +277,6 @@ static int send_client_reply_packet(MCPVIO_EXT *mpvio,
     mysql->options.extension->tls_allow_invalid_server_cert= 1;
   }
 
-  /* if server doesn't support SSL, we need to return an error */
-  if (mysql->options.use_ssl && !(mysql->server_capabilities & CLIENT_SSL))
-  {
-    my_set_error(mysql, CR_SSL_CONNECTION_ERROR, SQLSTATE_UNKNOWN,
-                        ER(CR_SSL_CONNECTION_ERROR),
-                        "SSL is required, but the server does not support it");
-    goto error;
-  }
-
   /* Remove options that server doesn't support */
   mysql->client_flag= mysql->client_flag &
                        (~(CLIENT_COMPRESS | CLIENT_ZSTD_COMPRESSION | CLIENT_SSL | CLIENT_PROTOCOL_41) 

From 484a959574e06801e3e85b15fa142c8f07b6658b Mon Sep 17 00:00:00 2001
From: Daniel Lenski <dlenski@amazon.com>
Date: Wed, 14 Jun 2023 12:25:39 -0700
Subject: [PATCH 3/7] Add comments about usage of TLS/SSL in connection
 establishment

All new code of the whole pull request, including one or several files
that are either new files or modified ones, are contributed under the
BSD-new license. I am contributing on behalf of my employer Amazon Web
Services, Inc.
---
 libmariadb/mariadb_lib.c | 11 +++++++++++
 plugins/auth/my_auth.c   | 18 ++++++++++++++++--
 2 files changed, 27 insertions(+), 2 deletions(-)

diff --git a/libmariadb/mariadb_lib.c b/libmariadb/mariadb_lib.c
index 7bff96912..bc505d5b2 100644
--- a/libmariadb/mariadb_lib.c
+++ b/libmariadb/mariadb_lib.c
@@ -1971,6 +1971,17 @@ MYSQL *mthd_my_real_connect(MYSQL *mysql, const char *host, const char *user,
 
   mysql->client_flag= client_flag;
 
+  /* Until run_plugin_auth has completed, the connection
+   * cannot have been secured with TLS/SSL.
+   *
+   * This means that any client which expects to use a
+   * TLS/SSL-secured connection SHOULD NOT trust any
+   * communication received from the server prior to this
+   * point as being genuine; nor should either the client
+   * or the server send any confidential information up
+   * to this point.
+   */
+
   if (run_plugin_auth(mysql, scramble_data, scramble_len,
                              scramble_plugin, db))
     goto error;
diff --git a/plugins/auth/my_auth.c b/plugins/auth/my_auth.c
index 34bdfa29b..3ed041bd2 100644
--- a/plugins/auth/my_auth.c
+++ b/plugins/auth/my_auth.c
@@ -355,8 +355,19 @@ static int send_client_reply_packet(MCPVIO_EXT *mpvio,
       (mysql->client_flag & CLIENT_SSL))
   {
     /*
-      Send mysql->client_flag, max_packet_size - unencrypted otherwise
-      the server does not know we want to do SSL
+      Send UNENCRYPTED "Login Request" packet with mysql->client_flag
+      and max_packet_size, but no username; without this, the server
+      does not know we want to switch to SSL/TLS
+
+      FIXME: Sending this packet is a very very VERY bad idea. It
+      contains the client's preferred charset and flags in plaintext;
+      this can be used for fingerprinting the client software version,
+      and probable geographic location.
+
+      This offers a glaring opportunity for pervasive attackers to
+      easily target, intercept, and exploit the client-server
+      connection (e.g. "MITM all connections from known-vulnerable
+      client versions originating from countries X, Y, and Z").
     */
     if (ma_net_write(net, (unsigned char *)buff, (size_t) (end-buff)) || ma_net_flush(net))
     {
@@ -366,6 +377,9 @@ static int send_client_reply_packet(MCPVIO_EXT *mpvio,
                           errno);
       goto error;
     }
+    /* This is where the socket is actually converted from a plain
+     * TCP/IP socket to a TLS/SSL-wrapped socket.
+     */
     if (ma_pvio_start_ssl(mysql->net.pvio))
       goto error;
     if (mysql->net.tls_self_signed_error &&

From 188873c324129b607059152bcbc82f5e5021c03e Mon Sep 17 00:00:00 2001
From: Daniel Lenski <dlenski@amazon.com>
Date: Wed, 3 Apr 2024 15:48:39 -0700
Subject: [PATCH 4/7] Consider Unix socket, shared memory, and named pipe
 connections to be "as secure as TLS" in terms of encryption

All new code of the whole pull request, including one or several files
that are either new files or modified ones, are contributed under the
BSD-new license. I am contributing on behalf of my employer Amazon Web
Services, Inc.
---
 libmariadb/mariadb_lib.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/libmariadb/mariadb_lib.c b/libmariadb/mariadb_lib.c
index bc505d5b2..fe5ffe089 100644
--- a/libmariadb/mariadb_lib.c
+++ b/libmariadb/mariadb_lib.c
@@ -1946,7 +1946,18 @@ MYSQL *mthd_my_real_connect(MYSQL *mysql, const char *host, const char *user,
   /* We now know the server's capabilities. If the client wants TLS/SSL,
    * but the server doesn't support it, we should immediately abort.
    */
-  if (mysql->options.use_ssl && !(mysql->server_capabilities & CLIENT_SSL))
+  if (mysql->options.use_ssl && !(mysql->server_capabilities & CLIENT_SSL)
+      /* Unix socket, shared memory, and named pipe are considered secure
+       * even if client requires SSL and server doesn't support it,
+       * because they do not involve off-host network traffic.
+       */
+#ifndef _WIN32
+      && net->pvio->type != PVIO_TYPE_UNIXSOCKET
+      && net->pvio->type != PVIO_TYPE_SHAREDMEM
+#else
+      && net->pvio->type != PVIO_TYPE_NAMEDPIPE
+#endif
+  )
   {
     SET_CLIENT_ERROR(mysql, CR_SSL_CONNECTION_ERROR, SQLSTATE_UNKNOWN,
                      "Client requires TLS/SSL, but the server does not support it");

From 58a682d69f88272ff77d50da1aa870146d3cf13f Mon Sep 17 00:00:00 2001
From: Daniel Lenski <dlenski@amazon.com>
Date: Mon, 19 Jun 2023 17:09:54 -0700
Subject: [PATCH 5/7] [MDEV-34009] Follow instant failover error packets

If the server responds to our initial connection with an error packet having
error number `ER_INSTANT_FAILOVER`, and an error string formatted as
`|Human-readable message|host[:port]`, then redirect accordingly.

This redirection is performed in `mthd_my_real_connect`, which appears to be
the most appropriate location given that it already contains code for
retrying/repeating a new server connection.

All new code of the whole pull request, including one or several files
that are either new files or modified ones, are contributed under the
BSD-new license. I am contributing on behalf of my employer Amazon Web
Services, Inc.
---
 include/mysqld_error.h   |  3 ++-
 libmariadb/mariadb_lib.c | 24 ++++++++++++++++++++++++
 2 files changed, 26 insertions(+), 1 deletion(-)

diff --git a/include/mysqld_error.h b/include/mysqld_error.h
index 133654374..ed340233b 100644
--- a/include/mysqld_error.h
+++ b/include/mysqld_error.h
@@ -1236,5 +1236,6 @@
 #define ER_JSON_HISTOGRAM_PARSE_FAILED 4186
 #define ER_SF_OUT_INOUT_ARG_NOT_ALLOWED 4187
 #define ER_INCONSISTENT_SLAVE_TEMP_TABLE 4188
-#define ER_ERROR_LAST 4188
+#define ER_INSTANT_FAILOVER 4200
+#define ER_ERROR_LAST 4200
 #endif /* ER_ERROR_FIRST */
diff --git a/libmariadb/mariadb_lib.c b/libmariadb/mariadb_lib.c
index fe5ffe089..44a7cd741 100644
--- a/libmariadb/mariadb_lib.c
+++ b/libmariadb/mariadb_lib.c
@@ -1731,6 +1731,7 @@ MYSQL *mthd_my_real_connect(MYSQL *mysql, const char *host, const char *user,
   else
 #endif
   {
+  tcp_redirect:
     cinfo.unix_socket=0;				/* This is not used */
     if (!port)
       port=mysql_port;
@@ -1995,7 +1996,30 @@ MYSQL *mthd_my_real_connect(MYSQL *mysql, const char *host, const char *user,
 
   if (run_plugin_auth(mysql, scramble_data, scramble_len,
                              scramble_plugin, db))
+  {
+    if (mysql->net.last_errno == ER_INSTANT_FAILOVER)
+    {
+      char *p= mysql->net.last_error; /* Should look like '|message|host[:port]' */
+      if (p && p[0] == '|')
+        p= strchr(p + 1, '|') ? : NULL;
+      if (p && *++p) {
+        host= p;
+        p= strchr(p, ':') ? : NULL;
+        if (p) {
+          *p++ = '\0';
+          port= atoi(p);
+        }
+        else
+        {
+          /* Restore to the default port, rather than reusing our current one */
+          port= 0;
+        }
+        fprintf(stderr, "Got instant failover to '%s' (port %d)\n", host, port);
+        goto tcp_redirect;
+      }
+    }
     goto error;
+  }
 
   if (mysql->client_flag & CLIENT_COMPRESS ||
       mysql->client_flag & CLIENT_ZSTD_COMPRESSION)

From 997d24520a48c51c4dbe6b5425d700977dd9a00d Mon Sep 17 00:00:00 2001
From: Daniel Lenski <dlenski@amazon.com>
Date: Wed, 5 Jul 2023 12:24:12 -0700
Subject: [PATCH 6/7] Allow clients to disable server-initiated instant
 failovers

This adds a new boolean option, `follow-instant-failovers` (ENABLED by
default), to allow the client to disable server-initiated instant failover
and simply return an error message.

All new code of the whole pull request, including one or several files
that are either new files or modified ones, are contributed under the
BSD-new license. I am contributing on behalf of my employer Amazon Web
Services, Inc.
---
 include/mysql.h          |  9 ++++++++
 libmariadb/mariadb_lib.c | 45 ++++++++++++++++++++++++++--------------
 2 files changed, 38 insertions(+), 16 deletions(-)

diff --git a/include/mysql.h b/include/mysql.h
index 1d30486a9..3408958a3 100644
--- a/include/mysql.h
+++ b/include/mysql.h
@@ -228,6 +228,8 @@ extern const char *SQLSTATE_UNKNOWN;
     /* MariaDB specific */
     MYSQL_PROGRESS_CALLBACK=5999,
     MYSQL_OPT_NONBLOCK,
+    MARIADB_OPT_FOLLOW_INSTANT_FAILOVERS,
+
     /* MariaDB Connector/C specific */
     MYSQL_DATABASE_DRIVER=7000,
     MARIADB_OPT_SSL_FP,             /* deprecated, use MARIADB_OPT_TLS_PEER_FP instead */
@@ -340,12 +342,19 @@ struct st_mysql_options {
     char *bind_address;
     my_bool secure_auth;
     my_bool report_data_truncation;
+    my_bool follow_instant_failovers;
+
     /* function pointers for local infile support */
     int (*local_infile_init)(void **, const char *, void *);
     int (*local_infile_read)(void *, char *, unsigned int);
     void (*local_infile_end)(void *);
     int (*local_infile_error)(void *, char *, unsigned int);
     void *local_infile_userdata;
+
+    /* WHY are some options relegated to this "extension" structure?
+     * What is the logic for distinguishing the "main" options from
+     * the extension options?
+     */
     struct st_mysql_options_extension *extension;
 };
 
diff --git a/libmariadb/mariadb_lib.c b/libmariadb/mariadb_lib.c
index 44a7cd741..afcf4f8cb 100644
--- a/libmariadb/mariadb_lib.c
+++ b/libmariadb/mariadb_lib.c
@@ -668,6 +668,7 @@ struct st_default_options mariadb_defaults[] =
   {{MYSQL_SET_CHARSET_NAME}, MARIADB_OPTION_STR, "default-character-set"},
   {{MARIADB_OPT_INTERACTIVE}, MARIADB_OPTION_NONE, "interactive-timeout"},
   {{MYSQL_OPT_CONNECT_TIMEOUT}, MARIADB_OPTION_INT, "connect-timeout"},
+  {{MARIADB_OPT_FOLLOW_INSTANT_FAILOVERS}, MARIADB_OPTION_BOOL, "follow-instant-failovers"},
   {{MYSQL_OPT_LOCAL_INFILE}, MARIADB_OPTION_BOOL, "local-infile"},
   {{0}, 0 ,"disable-local-infile",},
   {{MYSQL_OPT_SSL_CIPHER}, MARIADB_OPTION_STR, "ssl-cipher"},
@@ -1304,6 +1305,7 @@ mysql_init(MYSQL *mysql)
     goto error;
   mysql->options.report_data_truncation= 1;
   mysql->options.connect_timeout=CONNECT_TIMEOUT;
+  mysql->options.follow_instant_failovers= TRUE;
   mysql->charset= mysql_find_charset_name(MARIADB_DEFAULT_CHARSET);
   mysql->methods= &MARIADB_DEFAULT_METHODS;
   strcpy(mysql->net.sqlstate, "00000");
@@ -1999,23 +2001,31 @@ MYSQL *mthd_my_real_connect(MYSQL *mysql, const char *host, const char *user,
   {
     if (mysql->net.last_errno == ER_INSTANT_FAILOVER)
     {
-      char *p= mysql->net.last_error; /* Should look like '|message|host[:port]' */
-      if (p && p[0] == '|')
-        p= strchr(p + 1, '|') ? : NULL;
-      if (p && *++p) {
-        host= p;
-        p= strchr(p, ':') ? : NULL;
-        if (p) {
-          *p++ = '\0';
-          port= atoi(p);
-        }
-        else
-        {
-          /* Restore to the default port, rather than reusing our current one */
-          port= 0;
-        }
+      if (!mysql->options.follow_instant_failovers)
+      {
+        /* Client has disabled instant failover. Fall through and treat this
+         * as a "normal" error. */
+      }
+      else
+      {
+        char *p= mysql->net.last_error; /* Should look like '|message|host[:port]' */
+        if (p && p[0] == '|')
+          p= strchr(p + 1, '|') ? : NULL;
+        if (p && *++p) {
+          host= p;
+          p= strchr(p, ':') ? : NULL;
+          if (p) {
+            *p++ = '\0';
+            port= atoi(p);
+          }
+          else
+          {
+            /* Restore to the default port, rather than reusing our current one */
+            port= 0;
+          }
         fprintf(stderr, "Got instant failover to '%s' (port %d)\n", host, port);
-        goto tcp_redirect;
+          goto tcp_redirect;
+        }
       }
     }
     goto error;
@@ -3526,6 +3536,9 @@ mysql_optionsv(MYSQL *mysql,enum mysql_option option, ...)
   case MYSQL_OPT_RECONNECT:
     mysql->options.reconnect= *(my_bool *)arg1;
     break;
+  case MARIADB_OPT_FOLLOW_INSTANT_FAILOVERS:
+    mysql->options.follow_instant_failovers= *(my_bool *)arg1;
+    break;
   case MYSQL_OPT_PROTOCOL:
     mysql->options.protocol= *((uint *)arg1);
     break;

From fe2dd634f8d618ef52fe31321d8416171e2383e6 Mon Sep 17 00:00:00 2001
From: Daniel Lenski <dlenski@amazon.com>
Date: Wed, 5 Jul 2023 12:24:12 -0700
Subject: [PATCH 7/7] Limit number of server-initiated instant failovers
 followed by client

Set the limit to 8 for now. This will prevent the client from getting stuck
in redirect loops.

All new code of the whole pull request, including one or several files
that are either new files or modified ones, are contributed under the
BSD-new license. I am contributing on behalf of my employer Amazon Web
Services, Inc.
---
 libmariadb/mariadb_lib.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/libmariadb/mariadb_lib.c b/libmariadb/mariadb_lib.c
index afcf4f8cb..955c43137 100644
--- a/libmariadb/mariadb_lib.c
+++ b/libmariadb/mariadb_lib.c
@@ -85,6 +85,7 @@
 #define MA_RPL_VERSION_HACK "5.5.5-"
 
 #define CHARSET_NAME_LEN 64
+#define INSTANT_FAILOVER_LIMIT 8
 
 #undef max_allowed_packet
 #undef net_buffer_length
@@ -1593,7 +1594,7 @@ MYSQL *mthd_my_real_connect(MYSQL *mysql, const char *host, const char *user,
   my_bool is_multi= 0;
   char *host_copy= NULL;
   struct st_host *host_list= NULL;
-  int connect_attempts= 0;
+  int connect_attempts= 0, instant_failovers=0;
 
   if (!mysql->methods)
     mysql->methods= &MARIADB_DEFAULT_METHODS;
@@ -2006,6 +2007,13 @@ MYSQL *mthd_my_real_connect(MYSQL *mysql, const char *host, const char *user,
         /* Client has disabled instant failover. Fall through and treat this
          * as a "normal" error. */
       }
+      else if (instant_failovers >= INSTANT_FAILOVER_LIMIT)
+      {
+        /* Too many instant failovers */
+        my_set_error(mysql, ER_INSTANT_FAILOVER, SQLSTATE_UNKNOWN,
+                     "Too many instant failovers (>= %d)",
+                     INSTANT_FAILOVER_LIMIT);
+      }
       else
       {
         char *p= mysql->net.last_error; /* Should look like '|message|host[:port]' */
@@ -2024,6 +2032,7 @@ MYSQL *mthd_my_real_connect(MYSQL *mysql, const char *host, const char *user,
             port= 0;
           }
         fprintf(stderr, "Got instant failover to '%s' (port %d)\n", host, port);
+          ++instant_failovers;
           goto tcp_redirect;
         }
       }
